DOCS-#3850: Elaborate on modin.core.dataframe.base and .pandas (#3851)
modin-project/modin: docs/flow/modin/core/dataframe/base/index.rst

Purpose

> =======
> ``BaseDataframe`` serves the purpose of describing and defining the internal dataframe algebra.

Reviewer: ```suggestion
=======

``BaseDataframe`` serves the purpose of describing and defining the internal dataframe algebra.
```

> It is the core construction element which serves as a middle layer between high-level pandas-imitating API

Reviewer: I do not quite get this paragraph. You are saying here that base dataframe is a middle layer between API layer and QC layer whereas QC layer is actually a middle layer between API layer and base/core dataframe layer.

Author: Huh, now you wrote it I agree... I might've been fooled by some other diagram I have interpreted incorrectly.

and lower-level query compiler. ``BaseDataframe`` is the interface which implementations actually translate the dataframe algebra calls to queries to the underlying compiler.

The purpose of such translation is to reduce the vast amount of public pandas API to something smaller but sufficient.
Query compiler level could reduce the API even further depending on the implementation.

> Base dataframe and axis partitions are the interfaces that must be implemented by any backend that wants to be plugged in Modin.

Reviewer: I wish we would add a base partition manager and a base partition classes that just declare interfaces in near future.

Author: This is out of scope of this PR 🙃

Reviewer: Definitely

Reviewer: I was not able to leave a suggesting here so replace `backend` to `execution backend`/`execution system` please. We are trying to avoid a single `backend` word since the recent restructure.

> These classes are mostly abstract, however very simple and generic enough methods like

Reviewer: I would make base classes just declare interfaces.

Author: again, out of scope of docs PR 😉

Reviewer: yep

Purpose
=======

> ``BaseDataframe`` serves the purpose of describing and defining the internal dataframe algebra.

Reviewer: ```suggestion
``BaseDataframe`` serves the purpose of describing and defining the :doc:`Core Dataframe Algebra </flow/modin/core/dataframe/algebra>`.
```

> It is the core construction element which serves as the client for :doc:`Modin query compiler</flow/modin/core/storage_formats/base/query_compiler>` and which implementations are actually executing the queries from the compiler by invoking functions over partition(s).

Reviewer: ```suggestion
It is the core construction element which serves as the client for :doc:`Modin Query Compiler</flow/modin/core/storage_formats/base/query_compiler>` and which implementations are actually executing the queries from the compiler by invoking functions over partition(s).
```
